README.md
# MMM-rtsp-simple-server

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

This module serve [RTSPSimpleServer](https://github.com/aler9/rtsp-simple-server) instance and proxy multiple RTSP feeds to be inserted in MagicMirror through [Video.js](https://github.com/videojs/video.js) and [videojs-http-streaming (VHS)](https://github.com/videojs/http-streaming) to be used even within browsers not supporting video tags.

## Example

![](.github/example.png)

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-rtsp-simple-server',
            config: {
                updateInterval: 30000, // Default time to show next camera (in milliseconds)
                retryDelay: 5000, // Time to wait to refresh DOM when server and feeds are alive (in milliseconds)
                controls: false, // If video player should show its controls
                height: 350, // video player height
                width: 700, // video player width
                animationSpeed: 400, // Animation speed to update DOM
                sources: [], // sources list (rtsp urls to proxy. e.g rtsp://x.x.x.x:8554/live)
            }
        }
    ]
}
```

## Configuration options

| Option           | Default | Description                                                                              |
|------------------|---------|------------------------------------------------------------------------------------------|
| `updateInterval` | `30000` | *Optional* Default time to show next camera (in milliseconds)                            |
| `retryDelay`     | `5000`  | *Optional* Time to wait to refresh DOM when server and feeds are alive (in milliseconds) |
| `controls`       | `false` | *Optional* If video player should show its controls                                      |
| `height`         | `350`   | *Optional* video player height                                                           |
| `width`          | `700`   | *Optional* video player width                                                            |
| `animationSpeed` | `400`   | *Optional* Animation speed to update DOM                                                 |
| `sources`        | `[]`    | *Required* sources list (rtsp urls to proxy. e.g rtsp://x.x.x.x:8554/live)               |


CHANGELOG.md
# MMM-rtsp-simple-server Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Released

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: "MMM-rtsp-simple-server"
 *
 * By Andrés Vanegas <ukab72106@gmail.com>
 * MIT Licensed.
 */

const NodeHelper = require("node_helper");
const Log = require("../../js/logger.js");
const { createProxyMiddleware } = require("http-proxy-middleware");
const axios = require("axios").default;
const nocache = require('nocache');
const { spawn } = require('child_process');
const yaml = require('js-yaml');
const fs = require('fs');

module.exports = NodeHelper.create({
	name: "MMM-rtsp-simple-server",
	logPrefix: "MMM-rtsp-simple-server :: ",
	rtspServerDefaults: {
		logLevel: 'info',
		logDestinations: ["file", "stdout"],
		logFile: __dirname + "/bin/rtsp-simple-server.log",
		readTimeout: "2s",
		writeTimeout: "2s",
		readBufferCount: 1024,
		api: "no",
		metrics: "no",
		pprof: "no",
		rtspDisable: "yes",
		rtmpDisable: "yes",
		hlsDisable: "no",
		hlsAddress: ":8888",
		hlsAlwaysRemux: "yes",
		hlsVariant: "lowLatency",
		hlsSegmentCount: 10,
		hlsSegmentDuration: "1s",
		hlsPartDuration: "200ms",
		hlsSegmentMaxSize: "50M",
		hlsAllowOrigin: "*",
		hlsEncryption: "yes",
		hlsServerKey: __dirname + "/bin/rtsp-key.pem",
		hlsServerCert: __dirname + "/bin/rtsp.pem",
		hlsTrustedProxies: [],
		paths: {}
	},
	rtspServer: null,
	index: 0,
	paths: {},
	sources: {},
	readyState: false,
	proxyReadyState: false,

	start: function () {
		this.config = null;
		this.readyState = false;
		Log.info(this.logPrefix + "Started");
		this.sendNotification("SET_MESSAGE", "LOADING");
		this.resetRtspServer();
		setInterval(() => this.sendNotification("UPDATE_SOURCES", this.sources), 1000);
	},

	cleanName: function (x) {
		return x.replace(/[^a-z0-9]+/ig, "_");
	},

	processConfig: function (config) {
		this.config = config;

		var receivedConfigSources = config.sources.filter((v, i, self) => self.indexOf(v) === i);
		var payloadSources = receivedConfigSources.map(this.cleanName);
		var currentSources = Object.keys(this.sources);
		var newSources = payloadSources.filter(x => !currentSources.includes(x));
		this.rtspServerDefaults.readTimeout = Math.max(2, Math.round(config.updateInterval / 4000)) + "s";
		this.rtspServerDefaults.writeTimeout = Math.max(2, Math.round(config.updateInterval / 4000)) + "s";

		if (newSources.length > 0) {
			receivedConfigSources.forEach(x => {
				const key = this.cleanName(x);
				this.rtspServerDefaults.paths[key] = {
					source: x,
					sourceProtocol: 'udp'
				};
				this.sources[key] = '/' + this.name + '/stream/' + key + '/index.m3u8';
			})
			Log.info(this.logPrefix + "Sources updated", this.sources)
			this.resetRtspServer();
		}
	},

	resetRtspServer: function () {
		var self = this;
		if (self.rtspServer !== null) {
			try { self.rtspServer.kill(); } catch (_) { }
		}
		this.readyState = false;
		fs.writeFile(__dirname + '/bin/rtsp-simple-server.yml', yaml.dump(
			this.rtspServerDefaults,
			{ noCompatMode: true }),
			function (err) {
				if (err) {
					setTimeout(function () {
						self.resetRtspServer();
					}, 1000);
				} else {
					self.rtspServer = spawn(
						__dirname + '/bin/rtsp-simple-server',
						[__dirname + '/bin/rtsp-simple-server.yml'],
						{
							stdio: 'inherit',
						}
					);
					self.readyState = self.proxyReadyState;
				}
			}
		);
	},

	sendNotification(notification, payload) {
		this.sendSocketNotification(this.name + "-" + notification, payload);
		if (notification === 'UPDATE_SOURCES') {
			this.sendSocketNotification(this.name + "-READY", this.readyState);
		}
	},

	socketNotificationReceived: function (notification, payload) {
		notification = notification.replace(this.name + "-", "");

		switch (notification) {
			case "SET_CONFIG":
				this.processConfig(payload);
				if (!this.proxyReadyState) {
					this.setProxy();
				} else {
					this.sendNotification("UPDATE_SOURCES", this.sources);
				}
				break;
		}
	},

	// this you can create extra routes for your module
	setProxy: function () {
		var self = this;
		this.expressApp.set("etag", false);
		this.expressApp.use("/" + this.name + "/stream/*",
			nocache(),
			createProxyMiddleware({
				target: "https://127.0.0.1:8888", // target host with the same base path
				// proxyTimeout: this.config.retryDelay,
				// timeout: this.config.retryDelay,
				changeOrigin: true, // needed for virtual hosted sites
				secure: false,
				hostRewrite: true,
				protocolRewrite: true,
				followRedirects: true,
				pathRewrite: function (path, _) {
					return path.replace(new RegExp("^/" + self.name + "/stream/"), "/");
				},
			})
		);
		this.proxyReadyState = true;
		this.readyState = this.proxyReadyState;
	},
});


MMM-rtsp-simple-server.js
/* global Module */

/* Magic Mirror
 * Module: MMM-rtsp-simple-server
 *
 * By Andrés Vanegas <ukab72106@gmail.com>
 * MIT Licensed.
 */
Module.register("MMM-rtsp-simple-server", {
	/**
	 * @member {Object} defaults - Defines the default config values.
	 * @property {int} updateInterval Default time to show next source (in milliseconds). Defaults to 30000.
	 * @property {int} retryDelay Time to wait to refresh DOM when server and feeds are alive (in milliseconds). Defaults to 5000.
	 * @property {boolean} controls If video player should show its controls. Defaults to false.
	 * @property {int} height video player height. Defaults to 350.
	 * @property {int} width video player width. Defaults to 700.
	 * @property {int} animationSpeed Animation speed to update DOM. Defaults to 400.
	 * @property {str[]} sources sources list (rtsp urls to proxy. e.g rtsp://x.x.x.x:8554/live).
	*/
	defaults: {
		updateInterval: 30000,
		retryDelay: 5000,
		controls: false,
		height: 350,
		width: 700,
		animationSpeed: 400,
		mode: 'rotate',
		sources: [],
	},
	logPrefix: "MMM-rtsp-simple-server :: ",

	// Required version of MagicMirror
	requiresVersion: "2.1.0",

	// Placeholders
	wrapper: null,
	message: null,
	messageWrapper: null,
	playerWrappers: [],
	players: [],
	sources: [],
	sourcesOrder: [],
	readyState: false,
	currentIndex: 0,

	// Overrides start method
	start: function () {
		this.config = {
			...this.defaults,
			...this.config,
		};
		for (var i in this.config.sources) {
			this.playerWrappers[i] = null;
			this.players[i] = null;
		}
		if (this.config.mode === 'rotate') {
			this.rotateSource();
		} else if (this.config.mode === 'tiles') {
			this.tileSources();
		}
		this.sendNotification("SET_CONFIG", {
			...this.config,
			__protocol: window.location.protocol,
			__port: window.location.port,
		});
	},

	cleanName: function (x) {
		return x.replace(/[^a-z0-9]+/ig, "_");
	},

	updateSources: function (sources) {
		const validSources = this.config.sources.map(this.cleanName);
		var currentSources = Object.keys(this.sources);
		var payloadSources = Object.keys(sources).filter(x => validSources.indexOf(x) !== -1);

		var removedSources = currentSources.filter(x => !payloadSources.includes(x));
		var newSources = payloadSources.filter(x => !currentSources.includes(x));

		if (removedSources.length + newSources.length > 0) {
			this.currentIndex = 0;
			for (const [key, value] of Object.entries(sources)) {
				if (validSources.indexOf(key) !== -1) {
					this.sources[key] = value;
				}
			}
			this.sourcesOrder = Object.keys(this.sources);
			Log.info(this.logPrefix + "Sources updated", this.sources)
			this.updateDom(this.config.animationSpeed);
		}
	},

	/**
	 * Notification send helper method
	 * @param {string} notification notification type
	 * @param {any} payload notification payload
	 */
	sendNotification(notification, payload) {
		this.sendSocketNotification(this.name + "-" + notification, payload);
	},

	// Override socket notification received method
	socketNotificationReceived: function (notification, payload) {
		switch (notification.replace(this.name + "-", "")) {
			case "UPDATE_SOURCES":
				this.updateSources(payload);
				break;
			case "READY":
				this.readyState = payload;
				break;
		}
	},

	rotateSourcesOrder: function () {
		this.sourcesOrder.unshift(this.sourcesOrder.pop());
	},

	tileSources: function () {
		var self = this;
		var nextWaitTime = 0;
		if (this.config.mode !== 'tiles') {
			this.updateDom(this.config.animationSpeed);
		} else if (!this.readyState) {
			this.message = this.translate("LOADING");
			this.updateDom(this.config.animationSpeed);
			nextWaitTime = this.config.retryDelay;
		} else if (Object.keys(this.sources).length == 0) {
			this.message = this.translate("NO_SOURCES");
			this.updateDom(this.config.animationSpeed);
			nextWaitTime = this.config.retryDelay;
		} else {
			this.message = null;
			if (!this.players.every((_, i) => {
				return this.elementReady('playerWrappers', i) &&
					this.playerWrappers[i].offsetParent !== null &&
					this.elementReady('players', i);
			})) {
				this.showPlayers();
				nextWaitTime = this.config.animationSpeed;
			} else {
				nextWaitTime = this.config.updateInterval;
			}
		}
		setTimeout(function () { self.tileSources(); }, Math.max(nextWaitTime, 1000));
	},

	/**
	 * Change current source and activate player
	 * @param {object} source source data to be show
	 */
	rotateSource: function () {
		var self = this;
		var nextWaitTime = 0;
		if (this.config.mode !== 'rotate') {
			this.updateDom(this.config.animationSpeed);
		} else if (!this.readyState) {
			this.message = this.translate("LOADING");
			this.updateDom(this.config.animationSpeed);
			nextWaitTime = this.config.retryDelay;
		} else if (Object.keys(this.sources) == 0) {
			this.message = this.translate("NO_SOURCES");
			this.updateDom(this.config.animationSpeed);
			nextWaitTime = this.config.retryDelay;
		} else {
			if (!this.elementReady('players', 0)) {
				this.showPlayers();
				nextWaitTime = this.config.animationSpeed;
			} else {
				this.message = null;
				const key = this.sourcesOrder[0];
				const source = this.sources[key];
				nextWaitTime = this.config.updateInterval;
				Log.log("Showing source " + key);
				// this.player.poster("/" + this.name + source.image_url);
				this.players[0].src({
					src: source,
					type: 'application/x-mpegURL',
				});
				this.players[0].play();
				this.rotateSourcesOrder();
			}
		}
		setTimeout(function () { self.rotateSource(); }, Math.max(nextWaitTime, 1000));
	},

	/**
	 * Clears an attribute from this instance
	 * @param {string} nodename attribute to be clear
	 */
	clearNode(nodename) {
		switch (nodename) {
			case 'players':
				for (const [i, p] of this.players.entries()) {
					if (p !== null) {
						try { p.dispose(); } catch (_) { }
					}
					this.players[i] = null;
				}
				break;
			case 'playerWrappers':
				for (const [i, p] of this.playerWrappers.entries()) {
					if (p !== null) {
						try { p.parentNode.removeChild(p) } catch (_) { }
					}
					this.playerWrappers[i] = null;
				}
				break;
			default:
				if (this[nodename] !== null) {
					try { this[nodename].parentNode.removeChild(this[nodename]) } catch (_) { }
				}
				if (this.hasOwnProperty(nodename)) {
					this[nodename] = null;
				}
		}
	},

	elementReady: function (attribute, index = null) {
		return this[attribute] !== null &&
			(index === null || this[attribute][index] !== null);
	},

	/**
	 * Show message in wrapper and hide the player
	 */
	showMessage() {
		this.clearNode('players');
		this.clearNode('playerWrappers');

		if (!this.elementReady('messageWrapper')) {
			this.messageWrapper = document.createElement("div");
			this.messageWrapper.classList.add("message-container");
			this.messageWrapper.style.width = this.config.width + "px";
			this.messageWrapper.style.height = this.config.height + "px";
			this.wrapper.appendChild(this.messageWrapper);
		}
		this.messageWrapper.innerHTML = this.message;
	},

	createPlayer(index) {
		var self = this;
		if (!this.elementReady('playerWrappers', index)) {
			var playerWrapper = document.createElement("video-js");
			playerWrapper.classList.add("player_" + this.name);
			playerWrapper.setAttribute("id", "player_" + this.identifier + "-0");
			this.playerWrappers[index] = playerWrapper;
			this.wrapper.appendChild(this.playerWrappers[index]);
		}

		if (!this.elementReady('players', index) &&
			this.elementReady('playerWrappers', index) &&
			this.playerWrappers[index].offsetParent !== null) {
			try {
				var options = {
					autoplay: false,
					controls: this.config.controls,
					muted: "muted",
					preload: "none",
					width: this.config.width,
					height: this.config.height,
					fluid: true,
					liveui: true,
					loadingSpinner: false,
					enableSourceset: true,
					html5: {
						vhs: {
							overrideNative: true,
							experimentalBufferBasedABR: true,
							experimentalLLHLS: true,
							nativeAudioTracks: false,
							nativeVideoTracks: false,
						},
					},
				}
				if (this.config.mode === 'tiles') {
					options.autoplay = true;
					options.sources = { src: Object.values(this.sources).at(index) };
				}
				const resetPlayer = () => {
					try { this.players[index].dispose(); } catch (e) { Log.error(e); }
					this.players[index] = null;
					try { this.playerWrappers[index].parentNode.removeChild(this.playerWrappers[index]) } catch (_) { }
					this.playerWrappers[index] = null;
					this.createPlayer(index);
				};
				const resetPlayerInterval = () => {
					(this.players[index].readyState() <= 2) ?
						resetPlayer() :
						null;
				};
				this.players[index] = videojs(this.playerWrappers[index], options);
				this.players[index].setTimeout(resetPlayerInterval, this.defaults.retryDelay);
				this.players[index].on('pause', resetPlayer);
				this.players[index].on('stalled', resetPlayer);
				this.players[index].on('error', resetPlayer);
			} catch (e) {
				if (this.players[index] !== null) {
					try { this.players[index].dispose(); } catch (e) { }
				}
				this.players[index] = null;
				Log.error(e);
			}
		}
	},

	/**
	 * Show player and hide the message
	 */
	showPlayers() {
		this.clearNode('messageWrapper');
		if (this.config.mode === "rotate") {
			this.createPlayer(0);
		} else if (this.config.mode === "tiles") {
			for (var i = 0; i < Object.values(this.sources).length; i++) {
				this.createPlayer(i);
			}
		}
	},

	// Override function to retrieve DOM elements
	getDom: function () {
		if (this.wrapper === null) {
			this.wrapper = document.createElement("div");
			this.wrapper.classList.add("wrapper_" + this.name);
			this.wrapper.style.width = this.config.width + "px";
			this.wrapper.style.height = this.config.height + "px";
		}

		if (this.message !== null) {
			this.showMessage();
		} else {
			this.showPlayers();
		}

		return this.wrapper;
	},

	// Load scripts
	getScripts: function () {
		const __lang = this.config.lang || this.language || "en";
		return [
			this.file("js/videojs.min.js"),
			this.file("js/videojs-errors.min.js"),
			this.file("js/lang/" + (__lang) + ".js"),
			this.file("js/videojs-http-streaming-sync-workers.js"),
			this.file("js/videojs-http-streaming.min.js"),
		];
	},

	// Load stylesheets
	getStyles: function () {
		return [
			this.file("css/video-js.css"),
			this.name + ".css",
		];
	},

	// Load translations files
	getTranslations: function () {
		//FIXME: This can be load a one file javascript definition
		return {
			en: "translations/en.json",
			es: "translations/es.json"
		};
	},
});
